events/datadog_events.go

```diff
@@ -0,0 +1,110 @@
+package events
+
+import (
+	"bytes"
+	"fmt"
+
+	log "github.com/mozilla-services/reaper/reaperlog"
+)
+
+// DatadogStatistics implements EventReporter encapsulates DatadogEvents, sends events to Datadog
+// uses godspeed, requires dd-agent running
+type DatadogEvents struct {
+	Datadog
+}
+
+// NewDatadogEvents returns a new instance of DatadogEvents
+func NewDatadogEvents(c *DatadogConfig) *DatadogEvents {
+	c.Name = "DatadogEvents"
+	return &DatadogEvents{Datadog{Config: c}}
+}
+
+// NewEvent is a method of EventReporter
+// NewEvent reports an event to Datadog
+func (e *DatadogEvents) NewEvent(title string, text string, fields map[string]string, tags []string) error {
+	if e.Config.DryRun {
+		if log.Extras() {
+			log.Info("DryRun: Not reporting %s", title)
+		}
+		return nil
+	}
+
+	g, err := e.godspeed()
+	if err != nil {
+		return err
+	}
+	err = g.Event(title, text, fields, tags)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// NewReapableEvent is a method of EventReporter
+// NewReapableEvent is shorthand for a NewEvent about a reapable resource
+func (e *DatadogEvents) NewReapableEvent(r Reapable, tags []string) error {
+	if e.Config.shouldTriggerFor(r) {
+		text, err := r.ReapableEventText()
+		if err != nil {
+			return err
+		}
+		err = e.NewEvent("Reapable resource discovered", text.String(), nil, append(tags, "id:%s", r.ReapableDescriptionTiny()))
+		if err != nil {
+			return fmt.Errorf("Error reporting Reapable event for %s: %s", r.ReapableDescriptionTiny(), err.Error())
+		}
+	}
+	return nil
+}
+
+// NewBatchReapableEvent is a method of EventReporter
+func (e *DatadogEvents) NewBatchReapableEvent(rs []Reapable, tags []string) error {
+	var triggering []Reapable
+	for _, r := range rs {
+		if e.Config.shouldTriggerFor(r) {
+			triggering = append(triggering, r)
+		}
+	}
+	// no events triggering
+	if len(triggering) == 0 {
+		return nil
+	}
+
+	log.Info("Sending batch Datadog events for %d reapables.", len(triggering))
+	// this is a bin packing problem
+	// we ignore its complexity because we don't care (that much)
+	for j := 0; j < len(triggering); {
+		var written int64
+		buffer := *bytes.NewBuffer(nil)
+		for moveOn := false; j < len(triggering) && !moveOn; {
+			text, err := triggering[j].ReapableEventTextShort()
+			if err != nil {
+				return err
+			}
+			size := int64(text.Len())
+
+			// if there is room
+			if size+written < 4500 {
+				// write it + a newline
+				n, err := buffer.ReadFrom(text)
+				// not counting this length, but we have padding
+				_, err = buffer.WriteString("\n")
+				if err != nil {
+					return err
+				}
+				written += n
+				// increment counter of written reapables
+				j++
+			} else {
+				// if we've written enough to the buffer, break the loop
+				moveOn = true
+			}
+		}
+		// send events in this buffer
+		err := e.NewEvent("Reapable resources discovered", buffer.String(), nil, tags)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
```

Comment:
How about something like this instead? It's a little shorter and doesn't require 2 loops and as many tracking variables.
```
func (e *DatadogEvents) NewBatchReapableEvent(rs []Reapable, tags []string) error {
    errorStrings := []string{}
    buffer := new(bytes.Buffer)
    for _, r := range rs {
        if !e.Config.shouldTriggerFor(r) {
            continue
        }
        text, err := r.ReapableEventTextShort()
        if err != nil {
            errorStrings = append(errorStrings, fmt.Sprintf("ReapableEventText: %v", err))
            continue
        }
        if text.Len() > 4500 {
            text.Truncate(4499)
        }
        if text.Len()+buffer.Len() > 4500 {
            // send events in this buffer
            err := e.NewEvent("Reapable resources discovered", buffer.String(), nil, tags)
            if err != nil {
                errorStrings = append(errorStrings, fmt.Sprintf("NewEvent: %v", err))
            }
            buffer.Reset()
        }
        buffer.ReadFrom(text)
        buffer.WriteByte('\n')
    }

    // Flush remaining buffer
    if buffer.Len() > 0 {
        // send events in this buffer
        err := e.NewEvent("Reapable resources discovered", buffer.String(), nil, tags)
        if err != nil {
            errorStrings = append(errorStrings, fmt.Sprintf("NewEvent: %v", err))
        }
    }
    if len(errorStrings) > 0 {
        return errors.New(strings.Join(errorStrings, "\n"))
    }
    return nil
}
```